### PR TITLE
[JUJU-4342] Upgrade provider server protocol v5 to v6

### PR DESCRIPTION
**Pull request by cderici**
## Description

This upgrades the protocol version from 5 to 6, following [the translation guidelines](https://developer.hashicorp.com/terraform/plugin/mux/translating-protocol-version-5-to-6#testing-implementation).

Also upgrades the provider factories we use in the tests of resources that use the provider framework.

This is not necessarily related to the sdk2 -> framework migration, so things should work ok as it is, as the guidelines say:

```
Upgrading provider servers from protocol version 5 to protocol version 6 has no provider code requirements.
```

However, I tested the transition manually (`terraform init -upgrade`) with the following plan, as well as tested the code in #248 using this and seems to be working well. More manual tests are required for this, which is why I'm keeping it as a `WIP` for now.

```terraform
terraform {
  required_providers {
    juju = {
      version = "0.5.1"
      source  = "juju/juju"
    }
  }
}

provider "juju" {}

resource "juju_model" "development" {
  name = "test-upgrade"
}

resource "juju_application" "mysql" {
  name  = "mysql"
  model = juju_model.development.name
  charm {
    name   = "mysql"
    series = "jammy"
    channel = "8.0/stable"
  }

  units     = 1
  placement = juju_machine.my-machine.machine_id

  depends_on = [juju_machine.my-machine]
}

resource "juju_application" "wordpress" {
  name = "wordpress"

  model = juju_model.development.name

  units = 3
  charm {
    name = "wordpress"
    series = "xenial"
    channel = "stable"
  }
}

resource "juju_integration" "mysql_to_wp" {
  model = juju_model.development.name

  application {
    name     = juju_application.wordpress.name
  }

  application {
    name     = juju_application.mysql.name
  }
}

resource "juju_machine" "my-machine" {
  model = juju_model.development.name
  name  = "new-machine-auto-provision"
  series = "jammy"

  depends_on = [juju_model.development]
}

```

### CODE REVIEW

File: internal/provider/provider_test.go [hunk 1]
```diff
@@ -7,29 +7,54 @@ import (
...
 
 	frameworkprovider "github.com/hashicorp/terraform-plugin-framework/provider"
+	"github.com/hashicorp/terraform-plugin-framework/providerserver"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-mux/tf5to6server"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 	"github.com/stretchr/testify/assert"
+
```
Comment (hmlanigan) [inline internal/provider/provider_test.go:17]: Only 2 stanzas of imports required for this file, pls remove the new line.

**Review by hmlanigan — COMMENTED** [review]
A minor nit, but the tests are all failing.

File: internal/provider/resource_access_model_test.go [hunk 2]
```diff
@@ -52,7 +52,78 @@ func TestAcc_ResourceAccessModel_Basic(t *testing.T) {
...
+	modelName := "testing"
+	access := "write"
+	accessFail := "bogus"
+
+	resourceName := "juju_access_model.test"
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() { testAccPreCheck(t) },
+		ExternalProviders: map[string]resource.ExternalProvider{
+			"juju": {
+				VersionConstraint: "0.8.0",
```
Comment (hmlanigan) [inline internal/provider/resource_access_model_test.go:90]: "0.8.0" should be a constant so it's easy to change

**Review by hmlanigan — APPROVED** [review]
2 small changes, but lgtm

File: internal/provider/resource_user_test.go [hunk 3]
```diff
@@ -8,17 +8,62 @@ import (
...
 
-func TestAcc_ResourceUser_Basic(t *testing.T) {
+func TestAcc_ResourceUser_sdk2_framework_migrate(t *testing.T) {
+	userName := acctest.RandomWithPrefix("tfuser")
+	userPassword := acctest.RandomWithPrefix("tf-test-user")
+
+	resourceName := "juju_user.user"
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProviderFactories:        providerFactories,
```
Comment (hmlanigan) [inline internal/provider/resource_user_test.go:18]: You have duplicate factories... remove ProviderFactories

File: internal/provider/resource_user_test.go [hunk 4]
```diff
@@ -8,17 +8,62 @@ import (
...
+}
+
+func TestAcc_ResourceUser_Stable(t *testing.T) {
 	userName := acctest.RandomWithPrefix("tfuser")
 	userPassword := acctest.RandomWithPrefix("tf-test-user")
 
 	resourceName := "juju_user.user"
 	resource.Test(t, resource.TestCase{
 		PreCheck:          func() { testAccPreCheck(t) },
 		ProviderFactories: providerFactories,
```
Comment (hmlanigan) [inline internal/provider/resource_user_test.go:57]: only need ExternalProviders, remove ProviderFactories

File: version/version.go [hunk 5]
```diff
@@ -0,0 +1,3 @@
+package version
+
+const TerraformProviderJujuVersion = "0.8.0"
```
Comment (hmlanigan) [inline version/version.go:3]: We don't need a new package for this - it's test only. Can go in provider_test.go

Comment (cderici) [reply version/version.go:3]: I thought in the future we can do something with this, given that we have another constant sitting in the Makefile, maybe we can eliminate that one as well.

**Pull request merged**